torch/csrc/jit/ir/alias_analysis.cpp

```diff
@@ -41,41 +67,75 @@ class MutableTypePtrHelper {
   }
 
  private:
+  // This is a helper method for the container types in
+  // `getMutableTypeImpl`. Instead of simply calling `unshapedType`
+  // on the container type, we can save time by seeing if the type
+  // is already in `mutable_type_cache_`. This is kind of tricky
+  // because `List[Optional[T]]` should still be
+  // `List[Optional[Unshaped(T)]]`, while `mutable_type_cache_`
+  // would map `Optional[T]` to `T`
+  TypePtr unshapedContainerTypeWithCache(const TypePtr& type) {
+    if (!mutable_type_cache_) {
+      return unshapedType(type);
+    }
+    return type->withContained(
+        fmap(type->containedTypes(), [&](const TypePtr inner) {
+          auto maybe_type_mapping = mutable_type_cache_->find(inner);
+          if (maybe_type_mapping != mutable_type_cache_->end()) {
+            if (inner->kind() == maybe_type_mapping->first->kind()) {
+              return maybe_type_mapping->first;
+            }
+          }
+          return unshapedType(inner);
+        }));
+  }
+
   c10::optional<TypePtr> getMutableTypeImpl(const TypePtr& type) {
```

Comment:
In the previous implementation, we were returning a vector of types here? why was that changed ?

Reply:
In the current implementation, the signature is the same as it is here--`c10::optional<TypePtr> getMutableTypeImpl(const TypePtr& type)`. By not changing the signature, I was able to keep more of the original logic. Also, in most cases, we won't be returning multiple elements in the vector. Wouldn't returning a vector just cause lots of extra `if (vec.size() == ...` checks?

Reply:
I think some of those changes are needed, like here https://github.com/pytorch/pytorch/blob/6dea82a8edce728b8c4850b884bbb976a1a59c18/torch/csrc/jit/ir/alias_analysis.cpp#L1153. 

If we refine a `Union[List[int], Tensor]` to `Tensor` we want the refined Tensor Value* to point to the Union Value*, but that won't occur right now because they will be returning different Mutable Types. 

`getMutableType` could have always been `mapTypeToTypeSetWhichItCanAlias`, however because the only form of Unions previously we're `Union[T, None]` we always knew that the typeSet was size 1

Reply:
Resolved offline: my current implementation isn't wrong, but we will change `getMutableType` to return a `std::vector<TypePtr>` since that will make the API more intuitive